File: packages/tactical-engine/src/targeting/resolve-cube.js
import { blocksEffect } from "../validation/obstacle-rules";
import { canEffectReachAoE } from "./aoe-filter";
export function resolveCube(anchor, size, obstacles, edgeObstacles = []) {
    const result = [];
    for (let x = anchor.x; x < anchor.x + size; x += 1) {
        for (let y = anchor.y; y < anchor.y + size; y += 1) {
            const coordinate = { x, y };
            if (!blocksEffect(obstacles, coordinate) &&
                canEffectReachAoE(anchor, coordinate, obstacles, edgeObstacles)) {
                result.push(coordinate);
            }
        }
    }
    return result;
}

File: packages/tactical-engine-py/tactical_engine/targeting/resolve_cube.py
from shared_contracts import Coordinate, Obstacle
from ..validation.obstacle_rules import blocks_targeting


def resolve_cube(
    anchor: Coordinate,
    size: int,
    obstacles: list[Obstacle],
) -> list[Coordinate]:
    result: list[Coordinate] = []
    for x in range(anchor.x, anchor.x + size):
        for y in range(anchor.y, anchor.y + size):
            coordinate = Coordinate(x=x, y=y)
            if not blocks_targeting(obstacles, coordinate):
                result.append(coordinate)
    return result
